support/k8s/k8s-workload-registrar/mode-crd/controllers/endpoint_controller.go

```diff
@@ -0,0 +1,209 @@
+	}
+
+	svcName := getServiceDNSName(req.NamespacedName)
+	for _, subset := range endpoints.Subsets {
+		for _, address := range subset.Addresses {
+			if address.TargetRef == nil {
+				continue
+			}
+
+			// Get SPIFFE ID resource associated with this endpoint
+			spiffeIDList := spiffeidv1beta1.SpiffeIDList{}
+			labelSelector := labels.Set(map[string]string{
+				"podUid": string(address.TargetRef.UID),
+			})
+			err := e.List(e.c.Ctx, &spiffeIDList, &client.ListOptions{
+				LabelSelector: labelSelector.AsSelector(),
+			})
+			if err != nil {
+				e.c.Log.WithError(err).Error("Error getting spiffeid list")
+				return ctrl.Result{}, err
+			}
+
+			// If there are no SPIFFE ID resources associated with this endpoint, we may need to requeue
+			// Its possible this reconcile loop ran before the SPIFFE ID resource was generated
+			if len(spiffeIDList.Items) == 0 {
+				return ctrl.Result{
+					Requeue: e.requeue(address.TargetRef.Name, address.TargetRef.Namespace),
+				}, nil
+			}
+
+			// Iterate through the list of SPIFFE ID resources and update to add the DNS name
+			for _, spiffeID := range spiffeIDList.Items {
+				if !containsString(spiffeID.Spec.DnsNames[1:], svcName) {
+					spiffeID := spiffeID
+					spiffeID.Spec.DnsNames = append(spiffeID.Spec.DnsNames, svcName)
+					err := e.Update(e.c.Ctx, &spiffeID)
+					if err != nil {
+						return ctrl.Result{}, err
+					}
+
+					e.c.Log.WithFields(logrus.Fields{
+						"spiffeID": spiffeID.ObjectMeta.Name,
+						"dnsName":  svcName,
+					}).Info("Adding DNS name")
+				}
+			}
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// deleteExternalResources removes the service name from the list of DNS Names when the service is removed
+func (e *EndpointReconciler) deleteExternalResources(namespacedName types.NamespacedName) error {
+	svcName := getServiceDNSName(namespacedName)
+	spiffeIDList := spiffeidv1beta1.SpiffeIDList{}
+
+	err := e.List(e.c.Ctx, &spiffeIDList, &client.ListOptions{
+		Namespace: namespacedName.Namespace,
+	})
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			e.c.Log.WithFields(logrus.Fields{
+				"service": svcName,
+			}).WithError(err).Error("Failed to get list of SpiffeID CRDs")
+			return err
+		}
+
+		return nil
+	}
+
+	for _, spiffeID := range spiffeIDList.Items {
+		e.c.Log.WithFields(logrus.Fields{
+			"spiffeID": spiffeID.ObjectMeta.Name,
+		}).Info("Removing DNS names")
+
+		spiffeID := spiffeID
+		spiffeID.Spec.DnsNames = removeStringIf(spiffeID.Spec.DnsNames, svcName)
+
+		if err := e.Update(e.c.Ctx, &spiffeID); err != nil {
+			e.c.Log.WithFields(logrus.Fields{
+				"spiffeID": spiffeID.ObjectMeta.Name,
+			}).WithError(err).Error("Unable to delete DNS names in SpiffeID CRD")
+			return err
+		}
+	}
+
+	return nil
+}
+
+// requeue determines if the reconcile needs to be requeued. If the controller has been configured with a
+// pod label/annotation and the pod has the label/annotation then yes. If the controller has not been
```

Comment:
What about the case where it's in ServiceAccount mode?

Reply:
That will requeue each time because a spiffe id is generated for each pod since each pod has a service account. its not optional like the pod label/annotation.